components/places/src/storage/history.rs

```diff
@@ -188,6 +188,144 @@ pub fn delete_place_by_guid(db: &impl ConnExt, guid: &SyncGuid) -> Result<()> {
+        &visits,
+        |(visit_id, _)| visit_id,
+        |chunk, _| -> Result<()> {
+            db.conn().execute(
+                &format!(
+                    "DELETE from moz_historyvisits WHERE id IN ({})",
+                    sql_support::repeat_sql_vars(chunk.len()),
+                ),
+                chunk,
+            )?;
+            Ok(())
+        },
+    )?;
+
+    // Find out which pages have been possibly orphaned and clean them up.
+    sql_support::each_chunk_mapped(
+        &visits,
+        |(_, place_id)| place_id.0,
+        |chunk, _| -> Result<()> {
+            let query = format!(
+                "SELECT id, -- url, url_hash, guid
+                    (foreign_count != 0) AS has_foreign,
+                    ((last_visit_date_local + last_visit_date_remote) != 0) as has_visits
+                FROM moz_places
+                WHERE id IN ({})",
+                sql_support::repeat_sql_vars(chunk.len()),
+            );
+
+            let mut stmt = db.conn().prepare(&query)?;
+            let page_results = stmt.query_and_then(chunk, PageToClean::from_row)?;
+            let pages: Vec<PageToClean> = page_results.collect::<Result<_>>()?;
+            cleanup_pages(db, &pages)
+        },
+    )?;
+
+    Ok(())
+}
+
+#[derive(Debug)]
+struct PageToClean {
+    id: RowId,
+    has_foreign: bool,
+    has_visits: bool,
+}
+
+impl PageToClean {
+    pub fn from_row(row: &Row) -> Result<Self> {
+        Ok(Self {
+            id: row.get_checked("id")?,
+            has_foreign: row.get_checked("has_foreign")?,
+            has_visits: row.get_checked("has_visits")?,
+        })
+    }
+}
+
+/// Clean up pages whose history has been modified, by either
+/// removing them entirely (if they are marked for removal,
+/// typically because all visits have been removed and there
+/// are no more foreign keys such as bookmarks) or updating
+/// their frecency.
+fn cleanup_pages(db: &impl ConnExt, pages: &[PageToClean]) -> Result<()> {
+    // desktop does this frecency work using a function in a single sql
+    // statement - we should see if we can do that too.
+    let frec_ids: Vec<RowId> = pages
+        .iter()
+        .filter(|&p| p.has_foreign || p.has_visits)
+        .map(|p| p.id)
+        .collect();
+    for id in frec_ids {
+        update_frecency(db.conn(), id, None)?;
+    }
+    let to_remove: Vec<&PageToClean> = pages
+        .into_iter()
+        .filter(|p| !p.has_foreign && !p.has_visits)
+        .collect();
+    // Like desktop, we do "AND foreign_count = 0 AND last_visit_date ISNULL"
+    // to creating orphans in case of async race conditions - in Desktop's
+    // case, it reads the pages before starting a write transaction, so that
+    // probably is possible. We don't currently do that, but might later, so
+    // we do it anyway.
+    let remove_ids: Vec<RowId> = to_remove.iter().map(|p| p.id).collect();
+    sql_support::each_chunk(&remove_ids, |chunk, _| -> Result<()> {
```

Comment:
This function doesn't actually need to use chunking as it is called by a chunking parent. I should make that change, maybe with an assertion that the number of pages is less than the max chunk size.

Reply:
In that case, it's collect() is also not necessary, as you can pass an IntoIter directly into execute